In [1]:
# Standard libraries
import sys
# Add your custom path
gems_tco_path = "/Users/joonwonlee/Documents/GEMS_TCO-1/src"
sys.path.append(gems_tco_path)
import logging
import argparse # Argument parsing

# Data manipulation and analysis
import pandas as pd
import numpy as np
import pickle
import torch
import torch.optim as optim
import copy                    # clone tensor
import time

# Custom imports
import GEMS_TCO
from GEMS_TCO import kernels
from GEMS_TCO import data_preprocess 
from GEMS_TCO import kernels 
from GEMS_TCO import orderings as _orderings 
from GEMS_TCO import load_data
from GEMS_TCO import alg_optimization, alg_opt_Encoder
from GEMS_TCO import configuration as config

from typing import Optional, List, Tuple
from pathlib import Path
import typer
import json
from json import JSONEncoder

from GEMS_TCO.data_loader import load_data2

In [2]:
from pathlib import Path
# Assuming 'config' and 'load_data' class are defined and imported elsewhere

# --- Parameters derived from your framework ---
v: float = 0.5
space: List[str] = ['2', '2']
days: List[str] = ['0', '31']
mm_cond_number: int = 20
# --- End of framework parameters ---

lat_lon_resolution = [int(s) for s in space]
days_s_e = [int(d) for d in days]
days_list = list(range(days_s_e[0], days_s_e[1]))

# These values were not in the framework, so they remain as set in your snippet
years = ['2024']
month_range = [7] 

# Assuming 'config' is available in your environment
output_path = input_path = Path(config.mac_estimates_day_path)

## load ozone data from amarel
data_load_instance = load_data2(config.mac_data_load_path)

# Call the function using the variables from the framework
df_map, ord_mm, nns_map = data_load_instance.load_maxmin_ordered_data_bymonthyear(
lat_lon_resolution=lat_lon_resolution, 
mm_cond_number=mm_cond_number,
years_=years, 
months_=month_range,
lat_range=[0.0, 5.0],      # <-- Add this
lon_range=[123.0, 133.0]   # <-- Add this
)

Subsetting data to lat: [0.0, 5.0], lon: [123.0, 133.0]


In [12]:
df_day_aggregated_list = []
df_day_map_list = []
for i in range(31):
    idx_for_datamap = [i*8, (i+1)*8]

    cur_map, cur_df = analysis_map_no_mm, agg_data_no_mm = data_load_instance.load_working_data_keep_ori(
    df_map, 
    idx_for_datamap, 
    ord_mm=None,  # or just omit it
    dtype=torch.float # or just omit it
)
    df_day_aggregated_list.append( cur_df )
    df_day_map_list.append( cur_map )

print(df_day_aggregated_list[0].shape)
print(df_day_aggregated_list[1])



torch.Size([36480, 6])
tensor([[4.9760e+00, 1.3298e+02, 2.6703e+02, 4.5000e+01, 4.9764e+00, 1.3299e+02],
        [4.9760e+00, 1.3286e+02, 2.5886e+02, 4.5000e+01, 4.9766e+00, 1.3286e+02],
        [4.9760e+00, 1.3273e+02, 2.7201e+02, 4.5000e+01, 4.9769e+00, 1.3273e+02],
        ...,
        [4.8000e-02, 1.2328e+02, 2.5685e+02, 5.2000e+01, 5.2389e-02, 1.2329e+02],
        [4.8000e-02, 1.2316e+02, 2.5635e+02, 5.2000e+01, 5.2416e-02, 1.2316e+02],
        [4.8000e-02, 1.2303e+02, 2.6741e+02, 5.2000e+01, 5.2566e-02, 1.2304e+02]])


# swap columns

In [6]:
df_day_aggregated_list = []
df_day_map_list = []
for i in range(31):
    idx_for_datamap = [i*8, (i+1)*8]

    cur_map, cur_df = analysis_map_no_mm, agg_data_no_mm = data_load_instance.load_working_data_keep_ori(
    df_map, 
    idx_for_datamap, 
    ord_mm=None,  # or just omit it
    dtype=torch.float # or just omit it 
    )
    cur_df[:,0], cur_df[:,1] = cur_df[:,4] , cur_df[:,5]

    for map in cur_map.values():
        map[:,0], map[:,1] = map[:,4] , map[:,5]
    df_day_aggregated_list.append( cur_df )
    df_day_map_list.append( cur_map )

print(df_day_aggregated_list[1])

tensor([[4.9764e+00, 1.3299e+02, 2.6703e+02, 4.5000e+01, 4.9764e+00, 1.3299e+02],
        [4.9766e+00, 1.3286e+02, 2.5886e+02, 4.5000e+01, 4.9766e+00, 1.3286e+02],
        [4.9769e+00, 1.3273e+02, 2.7201e+02, 4.5000e+01, 4.9769e+00, 1.3273e+02],
        ...,
        [5.2389e-02, 1.2329e+02, 2.5685e+02, 5.2000e+01, 5.2389e-02, 1.2329e+02],
        [5.2416e-02, 1.2316e+02, 2.5635e+02, 5.2000e+01, 5.2416e-02, 1.2316e+02],
        [5.2566e-02, 1.2304e+02, 2.6741e+02, 5.2000e+01, 5.2566e-02, 1.2304e+02]])


In [ ]:
instance1 = kernels.vecchia_experiment(0.5, df_day_map_list[0], df_day_aggregated_list[0], nns_map, mm_cond_number, nheads=10)

a = [21.303, 1.307, 1.563, 0.022, -0.144, 0.198, 4.769]
#a = [30.2594, 0.665, 1.8981, 0.0, 0.1317, -0.0, 1.9785]
#a = [45.1402, 0.6299, 0.7308, -0.0003, -0.0151, 0.0, 7.8922]
#a = [21.7335, 1.2817, 1.5946, 0.042, -0.1241, 0.218, 4.8654]
#a = [20.453542336448137, 1.4506118600616982, 2.43096923637867, -0.03476556019978718, -0.1559262606484541, 0.1254833595232136, 3.938183829354925]
params = torch.tensor(a, dtype=torch.float64, requires_grad=True)

cov_map = instance1.cov_structure_saver(params, instance1.matern_cov_anisotropy_v05)  
instance1.vecchia_oct22( params, instance1.matern_cov_anisotropy_v05, cov_map )


# 15994.2207 15995.6299
# 64351   64359

tensor(64359.1719, grad_fn=<AddBackward0>)

In [8]:
v = 0.5 # smooth
mm_cond_number = 20
nheads = 300
lr = 0.02
step = 100
gamma_par = 0.3
epochs = 900

# fit just one hour

In [67]:
cur_df

tensor([[4.8780e+00, 1.3298e+02, 2.4773e+02, 7.4100e+02, 4.8780e+00, 1.3298e+02],
        [4.8784e+00, 1.3273e+02, 2.5376e+02, 7.4100e+02, 4.8784e+00, 1.3273e+02],
        [4.8788e+00, 1.3248e+02, 2.5099e+02, 7.4100e+02, 4.8788e+00, 1.3248e+02],
        ...,
        [1.2842e-01, 1.2367e+02, 2.5461e+02, 7.4800e+02, 1.2842e-01, 1.2367e+02],
        [1.2849e-01, 1.2342e+02, 2.5616e+02, 7.4800e+02, 1.2849e-01, 1.2342e+02],
        [1.2889e-01, 1.2317e+02, 2.5457e+02, 7.4800e+02, 1.2889e-01, 1.2317e+02]])

fit on center matching

In [13]:
# ------------------------------------
# Main Optimization Loop (Updated)
# ------------------------------------

# --- 1. Pre-load all data (based on reference code) ---
print("Pre-loading data for all days...")
df_day_aggregated_list = []
df_day_map_list = []
num_days_to_load = 31 # From reference code

for i in range(num_days_to_load):
    idx_for_datamap = [i*8, i*8+1]
    #idx_for_datamap = [i*8, (i+1)*8]    
    # Using the new load function from the reference code
    cur_map, cur_df = data_load_instance.load_working_data(
        df_map, 
        idx_for_datamap, 
        ord_mm=None,
        # Using float64 to ensure compatibility with kernels
        dtype=torch.float64 
    )
    df_day_aggregated_list.append( cur_df )
    df_day_map_list.append( cur_map )

print(f"Data loaded for {len(df_day_map_list)} days.")
# print(df_day_aggregated_list[0].shape) # From reference code

# --- 2. Run optimization loop over pre-loaded data ---

# This list is now just for iterating
#days_list = range(len(df_day_map_list)) 
days_list = [0]
for day in days_list:  
    
    # Get the pre-loaded data for this day
    analysis_data_map = df_day_map_list[day]
    aggregated_data = df_day_aggregated_list[day]

    #a = [21.303, 1.307, 1.563, 0.022, -0.144, 0.198, 4.769]
    a = [28.75, 0.98, 1.06, 0, 0, 0, 1.890]
    #a = [28.75, 0.98, 1.06, 0.036, -0.155, 0.179, 1.890]
    params = torch.tensor(a, dtype=torch.float64, requires_grad=True)
    
    # Calculate resolution for printing
    res_calc = (200 / lat_lon_resolution[0]) * (100 / lat_lon_resolution[0])
    print(f'\n--- Starting Day {day+1} (2024-07-{day+1}) ---')
    print(f'Data size per day: { res_calc }, smooth: {v}')
    print(f'mm_cond_number: {mm_cond_number},\ninitial parameters: \n {params.detach().numpy()}')
            
    # --- Data loading is now done *before* the loop ---

    # We need to define the device (though we aren't passing it anymore)
    device_str = 'cuda' if torch.cuda.is_available() else 'cpu'

    model_instance = kernels.model_fitting(
            smooth = v,
            input_map = analysis_data_map,
            aggregated_data = aggregated_data,
            nns_map = nns_map,
            mm_cond_number = mm_cond_number,
            nheads = nheads
            # device = device_str  <--- REMOVED: This was causing the TypeError
        )

    start_time = time.time()
    # Adjusted optimizer call based on expected return values (step size changed to step)
    optimizer, scheduler = model_instance.optimizer_fun(
        params, 
        lr=lr, 
        betas=(0.9, 0.8), 
        eps=1e-8, 
        step_size=step, # Using the 'step' variable here
        gamma=gamma_par  # Using gamma_par
    ) 

    # --- CRITICAL CORRECTION ---
    # 1. We no longer need to create a separate 'instance_map'.
    #    'model_instance' is already the correct instance.
    # 2. We do NOT pre-calculate 'cov_map'. The optimized training loop
    #    'run_vecc_may9' does this internally on each epoch
    #    to ensure the gradients are correct.
    # 3. We call 'run_vecc_may9' (the optimized loop) instead of 'run_vecc_grp9'.
    #    This version does not take 'cov_map' as an argument.
    
    # Calling the optimized 'run_vecc_may9'
    out, epoch_ran = model_instance.run_vecc_oct22(
        params, 
        optimizer,
        scheduler, 
        model_instance.matern_cov_anisotropy_v05, 
        epochs=epochs
    )
    # --- End Correction ---

    end_time = time.time()
    epoch_time = end_time - start_time
    print(f"Day {day+1} optimization finished in {epoch_time:.2f}s over {epoch_ran+1} epochs.")
    print(f"Day {day+1} final results: {out}")

print("\n--- All Days Processed ---")

Pre-loading data for all days...
Data loaded for 31 days.

--- Starting Day 1 (2024-07-1) ---
Data size per day: 5000.0, smooth: 0.5
mm_cond_number: 20,
initial parameters: 
 [28.75  0.98  1.06  0.    0.    0.    1.89]
Epoch 1, Gradients: [-9.93153032e-02  1.13098639e+01 -1.28471929e+01 -3.33955086e-13
 -5.68434189e-13  0.00000000e+00 -3.42754846e+00]
 Loss: 8564.823607970236, Parameters: [28.75  0.98  1.06  0.    0.    0.    1.89]
Epoch 51, Gradients: [-1.41217411e-01  3.25078944e-01  1.96046478e-01  6.96331881e-13
  1.30739863e-12  0.00000000e+00 -1.18153537e+00]
 Loss: 8357.356143726023, Parameters: [2.90119149e+01 1.17540012e+00 2.04530883e+00 2.09668554e-05
 1.46353349e-05 0.00000000e+00 2.90874622e+00]
Epoch 101, Gradients: [-1.02583159e-01  6.27706777e-02  2.08820949e-02  8.81072992e-13
  3.97903932e-13  0.00000000e+00 -1.76148904e-01]
 Loss: 8324.333578433401, Parameters: [ 3.00133200e+01  1.28192948e+00  2.22562651e+00  9.53212349e-06
 -2.31480545e-06  0.00000000e+00  3.137235

use swap order

In [11]:
# ------------------------------------
# Main Optimization Loop (Updated)
# ------------------------------------

# --- 1. Pre-load all data (based on reference code) ---
print("Pre-loading data for all days...")

df_day_aggregated_list = []
df_day_map_list = []
for i in range(31):
    #idx_for_datamap = [i*8, (i+1)*8]
    idx_for_datamap = [i*8, i*8 +1]

    cur_map, cur_df = analysis_map_no_mm, agg_data_no_mm = data_load_instance.load_working_data_keep_ori(
    df_map, 
    idx_for_datamap, 
    ord_mm=None,  # or just omit it
    dtype=torch.float # or just omit it 
    )
    cur_df[:,0], cur_df[:,1] = cur_df[:,4] , cur_df[:,5]

    for map in cur_map.values():
        map[:,0], map[:,1] = map[:,4] , map[:,5]
    df_day_aggregated_list.append( cur_df )
    df_day_map_list.append( cur_map )


print(f"Data loaded for {len(df_day_map_list)} days.")
# print(df_day_aggregated_list[0].shape) # From reference code

# --- 2. Run optimization loop over pre-loaded data ---

# This list is now just for iterating
#days_list = range(len(df_day_map_list)) 
days_list = [0]
for day in days_list:  
    
    # Get the pre-loaded data for this day
    analysis_data_map = df_day_map_list[day]
    aggregated_data = df_day_aggregated_list[day]

    #a = [21.303, 1.307, 1.563, 0.022, -0.144, 0.198, 4.769]
    a = [28.75, 0.98, 1.06, 0, 0, 0, 1.890]
    #a = [28.75, 0.98, 1.06, 0.036, -0.155, 0.179, 1.890]
    params = torch.tensor(a, dtype=torch.float64, requires_grad=True)
    
    # Calculate resolution for printing
    res_calc = (200 / lat_lon_resolution[0]) * (100 / lat_lon_resolution[0])
    print(f'\n--- Starting Day {day+1} (2024-07-{day+1}) ---')
    print(f'Data size per day: { res_calc }, smooth: {v}')
    print(f'mm_cond_number: {mm_cond_number},\ninitial parameters: \n {params.detach().numpy()}')
            
    # --- Data loading is now done *before* the loop ---

    # We need to define the device (though we aren't passing it anymore)
    device_str = 'cuda' if torch.cuda.is_available() else 'cpu'

    model_instance = kernels.model_fitting(
            smooth = v,
            input_map = analysis_data_map,
            aggregated_data = aggregated_data,
            nns_map = nns_map,
            mm_cond_number = mm_cond_number,
            nheads = nheads
            # device = device_str  <--- REMOVED: This was causing the TypeError
        )

    start_time = time.time()
    # Adjusted optimizer call based on expected return values (step size changed to step)
    optimizer, scheduler = model_instance.optimizer_fun(
        params, 
        lr=lr, 
        betas=(0.9, 0.8), 
        eps=1e-8, 
        step_size=step, # Using the 'step' variable here
        gamma=gamma_par  # Using gamma_par
    ) 

    # --- CRITICAL CORRECTION ---
    # 1. We no longer need to create a separate 'instance_map'.
    #    'model_instance' is already the correct instance.
    # 2. We do NOT pre-calculate 'cov_map'. The optimized training loop
    #    'run_vecc_may9' does this internally on each epoch
    #    to ensure the gradients are correct.
    # 3. We call 'run_vecc_may9' (the optimized loop) instead of 'run_vecc_grp9'.
    #    This version does not take 'cov_map' as an argument.
    
    # Calling the optimized 'run_vecc_may9'
    out, epoch_ran = model_instance.run_vecc_oct22(
        params, 
        optimizer,
        scheduler, 
        model_instance.matern_cov_anisotropy_v05, 
        epochs=epochs
    )
    # --- End Correction ---

    end_time = time.time()
    epoch_time = end_time - start_time
    print(f"Day {day+1} optimization finished in {epoch_time:.2f}s over {epoch_ran+1} epochs.")
    print(f"Day {day+1} final results: {out}")

print("\n--- All Days Processed ---")

Pre-loading data for all days...
Data loaded for 31 days.

--- Starting Day 1 (2024-07-1) ---
Data size per day: 5000.0, smooth: 0.5
mm_cond_number: 20,
initial parameters: 
 [28.75  0.98  1.06  0.    0.    0.    1.89]
Epoch 1, Gradients: [-1.05690956e-01  1.14465942e+01 -1.27860107e+01 -1.22070312e-04
  6.40869141e-04  0.00000000e+00 -3.60217286e+00]
 Loss: 8566.6640625, Parameters: [28.75  0.98  1.06  0.    0.    0.    1.89]
Epoch 51, Gradients: [-1.37584448e-01  3.43799591e-01  1.42578125e-01 -7.62939453e-04
 -2.44140625e-04  0.00000000e+00 -1.18694783e+00]
 Loss: 8360.35546875, Parameters: [29.04513558  1.1734433   2.03553031  0.07186149 -0.06086177  0.
  2.9143169 ]
Epoch 101, Gradients: [-0.10074802  0.04925156  0.0145874   0.00018311 -0.00015259  0.
 -0.15212309]
 Loss: 8325.1630859375, Parameters: [ 3.00445305e+01  1.28615078e+00  2.23070790e+00 -3.06708430e-03
 -3.17577525e-02  0.00000000e+00  3.15776994e+00]
Epoch 151, Gradients: [-0.09329809 -0.00375366 -0.0032959   0.000183

## Full likelihood

Vecchia Original
[29.921775534973218, 1.8597906958307489, 3.4954059102213106, 0.22512272633105687, -0.0648045117355326, 0.0, 3.3287024509849537] 2036
(data 1250) # 1 23641

Day 1 final results: [30.882481367893934, 1.3306287573904048, 2.3063941013205755, -0.05166757348340395, -0.020577100376109176, 0.0, 3.1918623924224208] 8314. (data 5000). 


Center matching
[29.867643794098317, 1.8170445928338301, 3.452060053856289, 9.346808116428221e-06, 1.3480203304274802e-05, 0.0, 3.278045234831132, 2038.106906954074]

[30.823365978489328, 1.325000558988864, 2.300043515398019, 5.191119207004892e-06, -9.752316964224805e-07, 0.0, 3.1756405728254657, 8313.954087906111]

# 1 23644


If I fit 8 hours on original space--> numerically unstable and cholesky decomposition fails after hitting non positive matrix


## Vecchia likleihood 8 hours using same parameters ()
a = [21.303, 1.307, 1.563, 0.022, -0.144, 0.198, 4.769]

original   15995.6299
center matching # 15994.2207 

Pre-loading data for all days...
Data loaded for 31 days.

--- Starting Day 1 (2024-07-1) ---
Data size per day: 1250.0, smooth: 0.5
mm_cond_number: 20,
initial parameters: 
 [28.75   0.98   1.06   0.036 -0.155  0.179  1.89 ]
Converged at epoch 535
Epoch 536,  
 vecc Parameters: [31.00915505  1.1916293   1.84680497  0.03927127 -0.15666047  0.12771191
  4.11954968]
FINAL STATE: Epoch 536, Loss: 15746.237437383388, 
 vecc Parameters: [31.009155050569746, 1.191629302554168, 1.846804972667287, 0.03927126531202711, -0.15666046904308123, 0.12771190935579058, 4.119549677954342]
Day 1 optimization finished in 822.07s over 536 epochs.
Day 1 final results: [31.009155050569746, 1.191629302554168, 1.846804972667287, 0.03927126531202711, -0.15666046904308123, 0.12771190935579058, 4.119549677954342, 15746.237437383388]

--- All Days Processed ---

Pre-loading data for all days...
Data loaded for 31 days.

--- Starting Day 1 (2024-07-1) ---
Data size per day: 1250.0, smooth: 0.5
mm_cond_number: 20,
initial parameters: 
 [28.75  0.98  1.06  0.    0.    0.    1.89]
Converged at epoch 332
Epoch 333,  
 vecc Parameters: [2.98676438e+01 1.81704459e+00 3.45206005e+00 9.34680812e-06
 1.34802033e-05 0.00000000e+00 3.27804523e+00]
FINAL STATE: Epoch 333, Loss: 2038.106906954074, 
 vecc Parameters: [29.867643794098317, 1.8170445928338301, 3.452060053856289, 9.346808116428221e-06, 1.3480203304274802e-05, 0.0, 3.278045234831132]
Day 1 optimization finished in 73.78s over 333 epochs.
Day 1 final results: [29.867643794098317, 1.8170445928338301, 3.452060053856289, 9.346808116428221e-06, 1.3480203304274802e-05, 0.0, 3.278045234831132, 2038.106906954074]

--- All Days Processed ---